### PR TITLE
Fix serialization restore for named tuples

### LINKED ISSUE #1429: Deserialized TrainState dosn't pass `_check_tree_and_avals` check triggered by jax control flow
### Problem you have encountered:
Training fails to run with restored `TrainState` when `jax.lax.cond` is in the loop. Specifically the `true_fun` and `false_fun` return type check fails because of having `optax._src...<TypeName>` vs `flax.serialization.<TypeName>` for updated and not-yet-updated branches respectively. 

### What you expected to happen:
Resume training without having any issues.

### Error message:
```
UnfilteredStackTrace: TypeError: true_fun and false_fun output must have same type structure, got PyTreeDef(CustomNode(<class 'flax.training.train_state.TrainState'>[(<bound method Module.apply of MLP(
    # attributes
    features = [4, 1]
)>, GradientTransformation(init=<function chain.<locals>.init_fn at 0x7f43f8875950>, update=<function chain.<locals>.update_fn at 0x7f43f8875a70>))], [*, CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}]), [CustomNode(namedtuple[<class 'optax._src.transform.ScaleByAdamState'>], [*, CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}]), CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}])]), CustomNode(namedtuple[<class 'flax.serialization.EmptyState'>], []), CustomNode(namedtuple[<class 'flax.serialization.EmptyState'>], [])]])) and PyTreeDef(CustomNode(<class 'flax.training.train_state.TrainState'>[(<bound method Module.apply of MLP(
    # attributes
    features = [4, 1]
)>, GradientTransformation(init=<function chain.<locals>.init_fn at 0x7f43f8875950>, update=<function chain.<locals>.update_fn at 0x7f43f8875a70>))], [*, CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}]), [CustomNode(namedtuple[<class 'flax.serialization.ScaleByAdamState'>], [*, CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}]), CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}])]), CustomNode(namedtuple[<class 'flax.serialization.EmptyState'>], []), CustomNode(namedtuple[<class 'flax.serialization.EmptyState'>], [])]])).

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------


The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)

<ipython-input-4-add6ba3b684b> in train_step(state, batch)
     23         return state
     24 
---> 25     new_state = jax.lax.cond(True, upd_step, no_upd_step, None)
     26     metrics = {"loss": loss}
     27 

TypeError: true_fun and false_fun output must have same type structure, got PyTreeDef(CustomNode(<class 'flax.training.train_state.TrainState'>[(<bound method Module.apply of MLP(
    # attributes
    features = [4, 1]
)>, GradientTransformation(init=<function chain.<locals>.init_fn at 0x7f43f8875950>, update=<function chain.<locals>.update_fn at 0x7f43f8875a70>))], [*, CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}]), [CustomNode(namedtuple[<class 'optax._src.transform.ScaleByAdamState'>], [*, CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}]), CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}])]), CustomNode(namedtuple[<class 'flax.serialization.EmptyState'>], []), CustomNode(namedtuple[<class 'flax.serialization.EmptyState'>], [])]])) and PyTreeDef(CustomNode(<class 'flax.training.train_state.TrainState'>[(<bound method Module.apply of MLP(
    # attributes
    features = [4, 1]
)>, GradientTransformation(init=<function chain.<locals>.init_fn at 0x7f43f8875950>, update=<function chain.<locals>.update_fn at 0x7f43f8875a70>))], [*, CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}]), [CustomNode(namedtuple[<class 'flax.serialization.ScaleByAdamState'>], [*, CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}]), CustomNode(<class 'flax.core.frozen_dict.FrozenDict'>[()], [{'params': {'Dense_0': {'bias': *, 'kernel': *}, 'Dense_1': {'bias': *, 'kernel': *}}}])]), CustomNode(namedtuple[<class 'flax.serialization.EmptyState'>], []), CustomNode(namedtuple[<class 'flax.serialization.EmptyState'>], [])]])).
```

### Steps to reproduce:
Linked [colab notebook](https://colab.research.google.com/drive/1J-mK1cWSunKCO9NgBA_Pb90PZcOTWgCi?usp=sharing) includes some workarounds. Below is code to reproduce:

```py
import jax
import jax.numpy as jnp
import flax
import flax.linen as nn
import optax

from typing import Sequence
from flax.training.train_state import TrainState
from flax.training.checkpoints import save_checkpoint, restore_checkpoint


rng = jax.random.PRNGKey(842)
rng, data_rng = jax.random.split(rng)
x = jnp.array([[x, x] for x in range(64)], dtype=jnp.float32)
y = jnp.sum(2*x + 1, axis=-1, keepdims=True)
x = x + jax.random.normal(data_rng, x.shape)
def data_gen():
    yield x, y
    
class MLP(nn.Module):
    features: Sequence[int]

    @nn.compact
    def __call__(self, x):
        for feat in self.features[:-1]:
            x = nn.relu(nn.Dense(feat)(x))
        x = nn.Dense(self.features[-1])(x)
        return x

model = MLP([4, 1])
params = model.init(jax.random.PRNGKey(0), x)

optimizer = optax.adamw(0.01)
optimizer = optax.MultiSteps(optimizer, 4)
state = TrainState.create(apply_fn=model.apply, params=params, tx=optimizer)


def compute_loss(params, batch):
    preds = state.apply_fn(params, batch[0])
    targs = batch[1]
    return jnp.mean((preds - targs)**2)

grad_fn = jax.value_and_grad(compute_loss)

def train_step(state, batch):
    def compute_loss(params):
        preds = state.apply_fn(params, batch[0])
        targs = batch[1]
        return jnp.mean((preds - targs)**2)
    grad_fn = jax.value_and_grad(compute_loss)
    
    loss, grad = grad_fn(state.params)

    new_state = state.apply_gradients(grads=grad)

    metrics = {"loss": loss}

    return new_state, metrics
train_step = jax.jit(train_step)

# train model, save checkpoint
for i in range(8):
    batch = next(data_gen())
    state, metrics = train_step(state, batch)
    print(metrics["loss"])
save_checkpoint('./_tmp/', state, 8, overwrite=True)

# restore checkopint, resume training - fails
state = restore_checkpoint('./_tmp/', state)
for i in range(8):
    batch = next(data_gen())
    state, metrics = train_step(state, batch)
    print(metrics["loss"])
```

### PR DESCRIPTION
**Pull request by jheek**
Serialization was creating a new named tuple class based on the fields in the state dict.
This results in a type error when comparing to the original or when comparing the structure of PyTrees.

Fixes #1429

### CODE REVIEW

**Comment by andsteing** [comment]
Thanks for the fix!

**Review by andsteing — APPROVED** [review]

File: flax/serialization.py [hunk 1]
```diff
@@ -125,27 +126,18 @@ def _restore_dict(xs, states):
 
 
 def _namedtuple_state_dict(nt):
-  return {'name': nt.__class__.__name__,
```
Comment (andsteing) [inline flax/serialization.py:128]: What's the motivation to also change the `state_dict` structure for fixing this?

**Review by andsteing — CHANGES_REQUESTED** [review]
Could you add a note to `changelog.md` to mention the change in the `state_dict` structure for named tuples?

File: flax/serialization.py (hunk 1 above)
Comment (jheek) [reply flax/serialization.py:128]: Having the structure like this is inconsistent with the rest of state_dict where we only store data (PyTree leaves) not structure. Also if you use the original NamedTuple you already have the name and fields so you don't have to store them in the state dict as well.

Comment (andsteing) [reply flax/serialization.py:128]: Great, thanks!

**Review by andsteing — APPROVED** [review]
Still maybe something worth to highlight in `changelog.md`?

**Review by andsteing — APPROVED** [review]
Great, thanks!

**Pull request merged**